stdlib/public/core/UInt128.swift

```diff
@@ -0,0 +1,411 @@
+    let (result, overflow) = a.multipliedReportingOverflow(by: b)
+    precondition(!overflow)
+    return result
+  }
+  
+  @inlinable
+  public static func *=(a: inout Self, b: Self) { a = a * b }
+  
+  @inlinable
+  public static func /(a: Self, b: Self) -> Self {
+    return a.dividedReportingOverflow(by: b).partialValue
+  }
+  
+  @inlinable
+  public static func /=(a: inout Self, b: Self) { a = a / b }
+  
+  @inlinable
+  public static func %(a: Self, b: Self) -> Self {
+    return a.remainderReportingOverflow(dividingBy: b).partialValue
+  }
+  
+  @inlinable
+  public static func %=(a: inout Self, b: Self) { a = a % b }
+}
+
+// MARK: - Numeric conformance
+@available(SwiftStdlib 6.0, *)
+extension UInt128: Numeric {
+  public typealias Magnitude = Self
+  
+  @inlinable
+  public var magnitude: Self { self }
+}
+
+// MARK: - BinaryInteger conformance
+@available(SwiftStdlib 6.0, *)
+extension UInt128: BinaryInteger {
+  
+  public struct Words {
+    @usableFromInline
+    let _value: UInt128
+    @usableFromInline
+    init(_value: UInt128) { self._value = _value }
+  }
+  
+  @inlinable
+  public var words: Words { Words(_value: self) }
+  
+  @inlinable
+  public static func &=(a: inout Self, b: Self) {
+    a._value = Builtin.and_Int128(a._value, b._value)
+  }
+  
+  @inlinable
+  public static func |=(a: inout Self, b: Self) {
+    a._value = Builtin.or_Int128(a._value, b._value)
+  }
+  
+  @inlinable
+  public static func ^=(a: inout Self, b: Self) {
+    a._value = Builtin.xor_Int128(a._value, b._value)
+  }
+  
+  @inlinable
+  public static func &>>=(a: inout Self, b: Self) {
+    let masked = b & 127
+    a._value = Builtin.lshr_Int128(a._value, masked._value)
+  }
+  
+  @inlinable
+  public static func &<<=(a: inout Self, b: Self) {
+    let masked = b & 127
+    a._value = Builtin.shl_Int128(a._value, masked._value)
+  }
+  
+  @inlinable
+  public var trailingZeroBitCount: Int {
+    low == 0 ? 64 + high.trailingZeroBitCount : low.trailingZeroBitCount
+  }
+}
+
+@available(SwiftStdlib 6.0, *)
+extension UInt128.Words: RandomAccessCollection {
+  public typealias Element = UInt
+  public typealias Index = Int
+  public typealias SubSequence = Slice<Self>
+  public typealias Indices = Range<Int>
+  
+  @inlinable public var count: Int { 128 / UInt.bitWidth }
+  @inlinable public var startIndex: Int { 0 }
+  @inlinable public var endIndex: Int { count }
+  @inlinable public var indices: Indices { startIndex ..< endIndex }
+  @inlinable public func index(after i: Int) -> Int { i + 1 }
+  @inlinable public func index(before i: Int) -> Int { i - 1 }
+  
+  public subscript(position: Int) -> UInt {
+    @inlinable
+    get {
+      precondition(position >= 0 && position < endIndex)
+      return withUnsafeBytes(of: _value) {
+#if _endian(little)
+        $0.assumingMemoryBound(to: UInt.self)[position]
+#else
+        $0.assumingMemoryBound(to: UInt.self)[count - 1 - position]
+#endif
+      }
```

Comment:
I believe you must use load(fromByteOffset:as:) to obey the abstract machine's definition of memory bound to T, since the memory at the location will be bound to UInt128 rather than UInt.

Reply:
Rolled into the latest update, thanks.